Implement SH1107 sleep/wake functionality

SAK917 opened the pull request:
Add `sleep()` and `wake()` methods and a `state` property to allow users to put the display into a power-saving mode.  Uses the `bus` object passed to the SH1107 class at initialization to write to the device.  Commands for entering and exiting sleep mode derived from the [SH1107 datasheet](https://www.displayfuture.com/Display/datasheet/controller/SH1107.pdf).

Tested on a Feather M4 Express and SH1107 FeatherWing OLED 128x64 display.

SAK917 committed:
Remove extra \x00 in DISPLAY ON of _INIT_SEQUENCE

SAK917 committed:
Implement sleep/wake states for display

Added:
  self._awake - stores current state of the display
  state - property that returns the current state
  sleep() - method that puts display to sleep
  wake() - method that wakes display from sleep

SAK917 committed:
Update .pre-commit-config.yaml

Received error during PR submission checks that the rev:
fields of the black and reuse-tool repos cannot be
mutable objects.  Changed rev from "latest" to the
current version numbers for each repo.

evaherrada commented:
@SAK917 Looks like I forgot to update some stuff on this library. I'll do that tomorrow and then merge it with your fork which will hopefully make your PR pass.

SAK917 committed:
Update copyright and licensing headers

PR check-in failed Pre-commit hooks test indicating
these files had missing copyright and licensing info.
Added headers as per @dherrada.

evaherrada commented:
@SAK917 Just heard you're working on adding the licenses. That'd be greatly appreciated! If you've got any questions, feel free to ping me

SAK917 committed:
Fix end-of-file error in .gitignore

SAK917 committed:
Reblacken examples using default black line length

SAK917 committed:
Address pylint errors and blacken

Addressed the following pylint errors:
   unused-import [13,1]
   redefined-builtin [16,1]
   wrong-import-order [16,1]
   wrong-import-order [17,1]

SAK917 committed:
Fix black formatting issue (different versions?)

SAK917 committed:
Fix docstring formatting for Sphinx

SAK917 commented:
> Just heard you're working on adding the licenses. That'd be greatly appreciated! If you've got any questions, feel free to ping me

@dherrada Happy to do what I can... thanks to a LOT of patience and help from @askpatrickw, I was able to muddle through all the changes the checks flagged.  Most were for the supporting files, sorry for all the commits!

evaherrada commented:
@SAK917 Great to see that it's passing!

askpatrickw reviewed (commented):
I reviewed the code and the changes look straight-forward.
I do not have this device so did not test.

I suggest **rebase and merge** when approved.

SAK917 committed:
Change name of display state property

Changed name of display state property from
"active" to "is_awake" to clarify it refers to the
display state.

SAK917 committed:
Change module reference in Usage Example (#3)

tannewt commented on adafruit_displayio_sh1107.py:
```diff
@@ -73,3 +72,36 @@ def __init__(self, bus, **kwargs):
         )
+        self.is_awake = True  # Display starts in active state (_INIT_SEQUENCE)
+
+    @property
+    def state(self):
+        """
+        The power state of the display. (read-only)
+
+        True if the display is active, False if in sleep mode.
+        """
+        return self.is_awake
```
I think `self.is_awake` is enough because it can be read directly. `state` as a property name is vague so `is_awake` is better for that too. You can put this comment after the `self.is_awake =` line above to have it documented in the docs. The only reason to do it as a property would be to make it read-only (and then the internal variable should start with `_` so it's clear it's internal.)

SAK917 replied:
Thank you Scott, I appreciate your input (and patience).  This is my first PR and I don't have much experience with classes, so it is a learning experience!

Upon reading your request, I realize I poorly executed the implementation of what I intended which is a read-only property.  I don't think we want `is_awake` to be writable since the `wake()` and `sleep()` methods handle updating the state, and if the user changes `is_awake` to the incorrect state then the `wake()` and `sleep()` methods will not work correctly.  However, I didn't like using `state` as a property name since it is vague as you stated above.

In your opinion would the following be more appropriate / correct to achieve the goal?

    self._is_awake = True  # Display starts in active state (_INIT_SEQUENCE)

    @property
    def is_awake(self):
        """
        The power state of the display. (read-only)
        True if the display is active, False if in sleep mode.
        """
        return self._is_awake

And please feel free to give me a direct recommendation on how best to do this... you won't step on my toes or hurt my feelings!  As I said previously, I am learning and definitely want to do things the way you feel best!

tannewt replied:
That's exactly what I was thinking! Nice work!

SAK917 replied:
Great, thanks for the feedback!  I will make the changes and push the update to the PR shortly.

tannewt reviewed (changes requested):
One API suggestion.

SAK917 committed:
Address API change request

Changed property name from state to is_awake
Made self._is_awake private

tannewt reviewed (approved):
Looks great! Thank you!
